examples/baseball.py

```diff
@@ -0,0 +1,273 @@
+from pyro.distributions.util import log_sum_exp
+from pyro.infer import EmpiricalMarginal
+from pyro.infer.abstract_infer import PosteriorPredictive
+from pyro.infer.mcmc import MCMC, NUTS
+
+"""
+Example has been adapted from [1]. It demonstrates how to do Bayesian inference using
+NUTS (or, HMC) in Pyro, and use of some common inference utilities.
+
+As in the Stan tutorial, this uses the small baseball dataset of Efron and Morris [2]
+to estimate players' batting average which is the fraction of times a player got a
+base hit out of the number of times they went up at bat.
+
+The dataset separates the initial 45 at-bats statistics from the remaining season.
+We use the hits data from the initial 45 at-bats to estimate the batting average
+for each player. We then use the remaining season's data to validate the predictions
+from our models.
+
+Three models are evaluated:
+ - Complete pooling model: The success probability of scoring a hit is shared
+     amongst all players.
+ - No pooling model: Each individual player's success probability is distinct and
+     there is no data sharing amongst players.
+ - Partial pooling model: A hierarchical model with partial data sharing.
+
+
+We recommend Radford Neal's tutorial on HMC ([3]) to users who would like to get a
+more comprehensive understanding of HMC and its variants, and to [4] for details on
+the No U-Turn Sampler, which provides an efficient and automated way (i.e. limited
+hyper-parameters) of running HMC on different problems.
+
+[1] Carpenter B. (2016), ["Hierarchical Partial Pooling for Repeated Binary Trials"]
+    (http://mc-stan.org/users/documentation/case-studies/pool-binary-trials.html).
+[2] Efron B., Morris C. (1975), "Data analysis using Stein's estimator and its
+    generalizations", J. Amer. Statist. Assoc., 70, 311-319.
+[3] Neal, R. (2012), "MCMC using Hamiltonian Dynamics",
+    (https://arxiv.org/pdf/1206.1901.pdf)
+[4] Hoffman, M. D. and Gelman, A. (2014), "The No-U-turn sampler: Adaptively setting
+    path lengths in Hamiltonian Monte Carlo", (https://arxiv.org/abs/1111.4246)
+"""
+
+logging.basicConfig(format='%(message)s', level=logging.INFO)
+# Enable validation checks
+pyro.enable_validation(True)
+pyro.set_rng_seed(1)
+DATA_URL = "http://www.swarthmore.edu/NatSci/peverso1/Sports%20Data/" + \
+           "JamesSteinData/Efron-Morris%20Baseball/EfronMorrisBB.txt"
+
+
+# ===================================
+#               MODELS
+# ===================================
+
+
+def fully_pooled(at_bats):
+    """
+    Number of hits in $K$ at bats for each player has a Binomial
+    distribution with a common probability of success, $\phi$.
+
+    :param (torch.Tensor) at_bats: Number of at bats for each player.
+    :return: Number of hits predicted by the model.
+    """
+    phi_prior = Uniform(at_bats.new_tensor(0), at_bats.new_tensor(1))
+    phi = pyro.sample("phi", phi_prior)
+    return pyro.sample("obs", Binomial(at_bats, phi))
```

Comment:
Do you mean subsampling?

Reply:
IIUC `iarange` would not help. First HMC is incompatible with subsampling. Second, without subsampling the only benefit provided by `iarange` is to declare independence. This declaration helps eliminate zero-expectation ELBO terms in SVI, but it is not used in HMC where we simply compute the total `log_prob` and its gradient.

Reply:
Good to know that! Now I somehow understand what you mentioned a while ago: use `iarange` instead of independent will reduce variance thanks to those eliminating jobs of iarange. Thanks @fritzo !